### PR TITLE
add `HLTFilter` template `HLTL1TMatchedJetsVBFFilter`

### PR DESCRIPTION
**Pull request by missirol**
#### PR description:

This PR adds new `HLTFilter`s which are meant to be used for an update of the VBF-parking triggers of the current HLT pp menu.

The new `HLTFilter`s are instantiations of the plugin template `HLTL1TMatchedJetsVBFFilter`, which serves as an improved version of the `EDProducer` template [`L1TJetsMatching`](https://github.com/cms-sw/cmssw/blob/6975683d2fb2d3c4f92f60a0266e49591dcdd042/RecoTauTag/HLTProducers/interface/L1TJetsMatching.h).

For further details, see [CMSHLT-2887](https://its.cern.ch/jira/browse/CMSHLT-2887).

Attn: @portalesHEP

#### PR validation:

Compared the results of the latest `GRun` menu before and after customising the relevant Paths with the introduction of instances of `HLTL1TMatchedPFJetsVBFFilter`. The trigger results [2] of the manual tests in [1] are as expected.

[1]
```bash
#!/bin/bash

jobLabel=hlt1
if [ ! -f "${jobLabel}".py ]; then
  hltGetConfiguration \
   /dev/CMSSW_13_0_0/GRun \
   --paths HLT_VBF_DiPFJet*,-HLT_VBF_DiPFJet*Detajj*,HLTriggerFinalPath \
   --max-events 1000 \
   --output minimal --no-prescale \
   --mc --globaltag 126X_mcRun3_2023_forPU65_v5 \
   --input /store/mc/Run3Winter23Digi/VBFHHto2B2Tau_CV-1_C2V-2_C3-1_TuneCP5_13p6TeV_madgraph-pythia8/GEN-SIM-RAW/126X_mcRun3_2023_forPU65_v1-v2/40000/a94d16bb-8df9-4eab-8c89-a5594aef6a48.root \
   --eras Run3 --l1-emulator FullMC --l1 L1Menu_Collisions2023_v1_3_0_xml \
   > "${jobLabel}".py
  cat <<EOF >> "${jobLabel}".py
process.options.numberOfThreads = 1
process.options.numberOfStreams = 0
EOF
  echo -e "\n=== ${jobLabel} ==="
  edmConfigDump "${jobLabel}".py > "${jobLabel}"_dump.py && \
   cmsRun "${jobLabel}".py &> "${jobLabel}".log && \
   mv output.root "${jobLabel}".root && \
   grep TrigRepor "${jobLabel}".log | head -20
else
  echo "Skipping ${jobLabel}"
fi

# customise using the new filter without changing the cuts in any HLT Paths (expect same exact trigger results)
jobLabel=hlt2
if [ ! -f "${jobLabel}".py ]; then
  cp hlt1.py "${jobLabel}".py
  cat <<EOF >> "${jobLabel}".py
from HLTrigger.JetMET.hltL1TMatchedPFJetsVBFFilter_cfi import hltL1TMatchedPFJetsVBFFilter as _hltL1TMatchedPFJetsVBFFilter

def _redefineLastFilterOfVBFParkingPaths(process, producerLabel, filterLabels):

    mod_p = getattr(process, producerLabel)

    for filterLabel in filterLabels:
        mod_f = getattr(process, filterLabel)

        setattr(process, filterLabel, _hltL1TMatchedPFJetsVBFFilter.clone(
            src = mod_p.JetSrc,
            maxJetDeltaR = mod_p.matchingR,
            l1tJetRefs = mod_p.L1JetTrigger,
            algorithm = mod_p.matchingMode,
            minPt1 = mod_p.pt1Min,
            minPt2 = mod_p.pt2Min,
            minPt3 = mod_p.pt3Min,
            minInvMass = mod_p.mjjMin,
            minNJets = mod_f.MinN,
            maxNJets = mod_f.MinN
        ))

    delattr(process, producerLabel)

    return process

_fooDict = {
    'hltL1PFJetsMatchingVBFinclMedium1050': ['hltL1PFJetCategoriesVBFinclMedium1050', 'hltL1PFJetCategoriesVBFinclMedium1050TripleJet'],
    'hltL1PFJetsMatchingVBFinclTight1050': ['hltL1PFJetCategoriesVBFinclTight1050', 'hltL1PFJetCategoriesVBFinclTight1050TripleJet'],
    'hltL1TPFJetsMatchingVBFDijetTight650': ['hltL1PFJetCategoriesVBFdijetTightQuadjet650', 'hltL1PFJetCategoriesVBFdijetTightFivejet650', 'hltL1PFJetCategoriesVBFdijetTightSixjet650'],
    'hltL1PFJetsMatchingVBFMETTight550': ['hltL1PFJetCategoriesVBFMETTight550', 'hltL1PFJetCategoriesVBFMETTight550Triplejet'],
    'hltL1PFJetsMatchingVBFMuTight650': ['hltL1PFJetCategoriesVBFMuTight650', 'hltL1PFJetCategoriesVBFMuTight650Triplejet'],
}

for key,val in _fooDict.items():
    process = _redefineLastFilterOfVBFParkingPaths(process, key, val)
EOF
  echo -e "\n=== ${jobLabel} ==="
  edmConfigDump "${jobLabel}".py > "${jobLabel}"_dump.py && \
   cmsRun "${jobLabel}".py &> "${jobLabel}".log && \
   mv output.root "${jobLabel}".root && \
   grep TrigRepor "${jobLabel}".log | head -20
else
  echo "Skipping ${jobLabel}"
fi

# customise fully (new filter, "merge" paths using maxNJets=-1),
# expect remaining Paths to the logical OR of the old ones (hlt1)
jobLabel=hlt3
if [ ! -f "${jobLabel}".py ]; then
  cp hlt1.py "${jobLabel}".py
  cat <<EOF >> "${jobLabel}".py
from HLTrigger.JetMET.hltL1TMatchedPFJetsVBFFilter_cfi import hltL1TMatchedPFJetsVBFFilter as _hltL1TMatchedPFJetsVBFFilter

def _redefineLastFilterOfVBFParkingPaths(process, producerLabel, filterLabels):

    mod_p = getattr(process, producerLabel)

    for filterLabel in filterLabels:
        mod_f = getattr(process, filterLabel)

        setattr(process, filterLabel, _hltL1TMatchedPFJetsVBFFilter.clone(
            src = mod_p.JetSrc,
            maxJetDeltaR = mod_p.matchingR,
            l1tJetRefs = mod_p.L1JetTrigger,
            algorithm = mod_p.matchingMode,
            minPt1 = mod_p.pt1Min,
            minPt2 = mod_p.pt2Min,
            minPt3 = mod_p.pt3Min,
            minInvMass = mod_p.mjjMin,
            minNJets = mod_f.MinN,
            maxNJets = -1
        ))

    delattr(process, producerLabel)

    return process

_fooDict = {
    'hltL1PFJetsMatchingVBFinclMedium1050': ['hltL1PFJetCategoriesVBFinclMedium1050'],
    'hltL1PFJetsMatchingVBFinclTight1050': ['hltL1PFJetCategoriesVBFinclTight1050'],
    'hltL1TPFJetsMatchingVBFDijetTight650': ['hltL1PFJetCategoriesVBFdijetTightQuadjet650'],
    'hltL1PFJetsMatchingVBFMETTight550': ['hltL1PFJetCategoriesVBFMETTight550'],
    'hltL1PFJetsMatchingVBFMuTight650': ['hltL1PFJetCategoriesVBFMuTight650'],
}

for key,val in _fooDict.items():
    process = _redefineLastFilterOfVBFParkingPaths(process, key, val)

for foo in [foo for foo in process.paths_() if '_TriplePFJet_v' in foo or 'FiveJet_v' in foo or 'SixJet_v' in foo]:
    delattr(process, foo)
EOF
  echo -e "\n=== ${jobLabel} ==="
  edmConfigDump "${jobLabel}".py > "${jobLabel}"_dump.py && \
   cmsRun "${jobLabel}".py &> "${jobLabel}".log && \
   mv output.root "${jobLabel}".root && \
   grep TrigRepor "${jobLabel}".log | head -20
else
  echo "Skipping ${jobLabel}"
fi
```

[2]
```
=== hlt1 ===
TrigReport ---------- Event  Summary ------------
TrigReport Events total = 1000 passed = 607 failed = 393
TrigReport ---------- Path   Summary ------------
TrigReport  Trig Bit#   Executed     Passed     Failed      Error Name
TrigReport     1    0       1000        318        682          0 HLT_VBF_DiPFJet110_40_Mjj1050_v1
TrigReport     1    1       1000         62        938          0 HLT_VBF_DiPFJet110_40_Mjj1050_TriplePFJet_v1
TrigReport     1    2       1000        289        711          0 HLT_VBF_DiPFJet125_45_Mjj1050_v1
TrigReport     1    3       1000         69        931          0 HLT_VBF_DiPFJet125_45_Mjj1050_TriplePFJet_v1
TrigReport     1    4       1000        379        621          0 HLT_VBF_DiPFJet75_45_Mjj650_DiPFJet60_JetMatchingQuadJet_v1
TrigReport     1    5       1000          0       1000          0 HLT_VBF_DiPFJet75_45_Mjj650_DiPFJet60_JetMatchingFiveJet_v1
TrigReport     1    6       1000          2        998          0 HLT_VBF_DiPFJet75_45_Mjj650_DiPFJet60_JetMatchingSixJet_v1
TrigReport     1    7       1000        341        659          0 HLT_VBF_DiPFJet80_45_Mjj550_PFMETNoMu85_v1
TrigReport     1    8       1000         25        975          0 HLT_VBF_DiPFJet80_45_Mjj550_PFMETNoMu85_TriplePFJet_v1
TrigReport     1    9       1000        172        828          0 HLT_VBF_DiPFJet95_45_Mjj650_Mu3_TrkIsoVVL_v1
TrigReport     1   10       1000         16        984          0 HLT_VBF_DiPFJet95_45_Mjj650_Mu3_TrkIsoVVL_TriplePFJet_v1
TrigReport     1   11       1000        264        736          0 HLT_VBF_DiPFJet45_Mjj550_MediumDeepTauPFTauHPS45_L2NN_eta2p1_v1
TrigReport     1   12       1000        183        817          0 HLT_VBF_DiPFJet50_Mjj550_Photon22_v1
TrigReport     1   13       1000         83        917          0 HLT_VBF_DiPFJet50_Mjj500_Ele22_eta2p1_WPTight_Gsf_v1
TrigReport     1   14       1000          0       1000          0 HLTriggerFinalPath
TrigReport -------End-Path   Summary ------------

=== hlt2 ===
TrigReport ---------- Event  Summary ------------
TrigReport Events total = 1000 passed = 607 failed = 393
TrigReport ---------- Path   Summary ------------
TrigReport  Trig Bit#   Executed     Passed     Failed      Error Name
TrigReport     1    0       1000        318        682          0 HLT_VBF_DiPFJet110_40_Mjj1050_v1
TrigReport     1    1       1000         62        938          0 HLT_VBF_DiPFJet110_40_Mjj1050_TriplePFJet_v1
TrigReport     1    2       1000        289        711          0 HLT_VBF_DiPFJet125_45_Mjj1050_v1
TrigReport     1    3       1000         69        931          0 HLT_VBF_DiPFJet125_45_Mjj1050_TriplePFJet_v1
TrigReport     1    4       1000        379        621          0 HLT_VBF_DiPFJet75_45_Mjj650_DiPFJet60_JetMatchingQuadJet_v1
TrigReport     1    5       1000          0       1000          0 HLT_VBF_DiPFJet75_45_Mjj650_DiPFJet60_JetMatchingFiveJet_v1
TrigReport     1    6       1000          2        998          0 HLT_VBF_DiPFJet75_45_Mjj650_DiPFJet60_JetMatchingSixJet_v1
TrigReport     1    7       1000        341        659          0 HLT_VBF_DiPFJet80_45_Mjj550_PFMETNoMu85_v1
TrigReport     1    8       1000         25        975          0 HLT_VBF_DiPFJet80_45_Mjj550_PFMETNoMu85_TriplePFJet_v1
TrigReport     1    9       1000        172        828          0 HLT_VBF_DiPFJet95_45_Mjj650_Mu3_TrkIsoVVL_v1
TrigReport     1   10       1000         16        984          0 HLT_VBF_DiPFJet95_45_Mjj650_Mu3_TrkIsoVVL_TriplePFJet_v1
TrigReport     1   11       1000        264        736          0 HLT_VBF_DiPFJet45_Mjj550_MediumDeepTauPFTauHPS45_L2NN_eta2p1_v1
TrigReport     1   12       1000        183        817          0 HLT_VBF_DiPFJet50_Mjj550_Photon22_v1
TrigReport     1   13       1000         83        917          0 HLT_VBF_DiPFJet50_Mjj500_Ele22_eta2p1_WPTight_Gsf_v1
TrigReport     1   14       1000          0       1000          0 HLTriggerFinalPath
TrigReport -------End-Path   Summary ------------

=== hlt3 ===
TrigReport ---------- Event  Summary ------------
TrigReport Events total = 1000 passed = 607 failed = 393
TrigReport ---------- Path   Summary ------------
TrigReport  Trig Bit#   Executed     Passed     Failed      Error Name
TrigReport     1    0       1000        380        620          0 HLT_VBF_DiPFJet110_40_Mjj1050_v1
TrigReport     1    1       1000        358        642          0 HLT_VBF_DiPFJet125_45_Mjj1050_v1
TrigReport     1    2       1000        381        619          0 HLT_VBF_DiPFJet75_45_Mjj650_DiPFJet60_JetMatchingQuadJet_v1
TrigReport     1    3       1000        366        634          0 HLT_VBF_DiPFJet80_45_Mjj550_PFMETNoMu85_v1
TrigReport     1    4       1000        188        812          0 HLT_VBF_DiPFJet95_45_Mjj650_Mu3_TrkIsoVVL_v1
TrigReport     1    5       1000        264        736          0 HLT_VBF_DiPFJet45_Mjj550_MediumDeepTauPFTauHPS45_L2NN_eta2p1_v1
TrigReport     1    6       1000        183        817          0 HLT_VBF_DiPFJet50_Mjj550_Photon22_v1
TrigReport     1    7       1000         83        917          0 HLT_VBF_DiPFJet50_Mjj500_Ele22_eta2p1_WPTight_Gsf_v1
TrigReport     1    8       1000          0       1000          0 HLTriggerFinalPath
TrigReport -------End-Path   Summary ------------
TrigReport  Trig Bit#   Executed     Passed     Failed      Error Name
TrigReport     0    0       1000       1000          0          0 @finalPath
TrigReport ---------- Modules in Path: HLT_VBF_DiPFJet110_40_Mjj1050_v1 ------------
TrigReport  Trig Bit#    Visited     Passed     Failed      Error Name
TrigReport     1    0       1000       1000          0          0 hltTriggerType
TrigReport     1    1       1000       1000          0          0 hltGtStage2Digis
```

#### If this PR is a backport, please specify the original PR and why you need to backport that PR. If this PR will be backported, please specify to which release cycle the backport is meant for:

`CMSSW_13_2_X`

Not needed for ppRef nor HIon, but useful to start the cleanup of the HLT pp menu ahead of 2024.

### CODE REVIEW

**Comment by missirol** [comment]
please test

**Comment by missirol** [comment]
please test

Minor technical changes in the last push. This should be the final version.

**Comment by perrotta** [comment]
+1

File: HLTrigger/JetMET/plugins/HLTL1TMatchedJetsVBFFilter.h [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+                                                   double const mjj) const {
+  // reset output jet indices
+  outputJetIndices.clear();
+
+  // find dijet pair with highest Mjj
+  int i1 = -1;
+  int i2 = -1;
+  double m2jj_max = -1;
+
+  for (unsigned int i = 0; i < jetIndices.size() - 1; i++) {
```
Comment (missirol) [inline HLTrigger/JetMET/plugins/HLTL1TMatchedJetsVBFFilter.h:174]: The use of `size() - 1` here is a bug, missed during the integration of this PR, spotted in https://github.com/cms-sw/cmssw/pull/43300#issuecomment-1817465701, and fixed in #43324 (14_0_X), #43325 (13_3_X) and #43326 (13_2_X)

**Pull request merged**